ambry-router/src/main/java/com.github.ambry.router/NonBlockingRouter.java

```diff
@@ -482,6 +507,100 @@ public void run() {
+   * same {@link OperationController} doing the put.
+   */
+  private class BackgroundDeleter extends OperationController {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    /**
+     * Instantiate the BackgroundDeleter
+     * @throws IOException if the associated {@link OperationController} throws one.
+     */
+    BackgroundDeleter() throws IOException {
+      super("backgroundDeleter");
+      super.putManager.close();
+    }
+
+    /**
+     * Put operations are disallowed in the BackgroundDeleter.
+     */
+    @Override
+    protected void putBlob(BlobProperties blobProperties, byte[] userMetadata, ReadableStreamChannel channel,
+        FutureResult<String> futureResult, Callback<String> callback) {
+      RouterException routerException =
+          new RouterException("Illegal attempt to put blob through backgroundDeleteOperationController",
+              RouterErrorCode.UnexpectedInternalError);
+      routerMetrics.operationDequeuingRate.mark();
+      routerMetrics.onPutBlobError(routerException);
+      operationCompleteCallback.completeOperation(futureResult, callback, null, routerException);
+    }
+
+    /**
+     * Initiate the deletes of the data chunks associated with this blobId, if this blob turns out to be a composite
+     * blob.
+     * @param blobId the blobId string associated with the possibly composite blob.
+     */
+    void initiateChunkDeletesIfAny(String blobId) {
+      FutureResult<List<StoreKey>> futureResult = new FutureResult<>();
+      Callback<List<StoreKey>> callback = new Callback<List<StoreKey>>() {
+        @Override
+        public void onCompletion(List<StoreKey> result, Exception exception) {
+          if (exception != null) {
+            logger.error("Encountered exception when attempting to get chunks of a possibly composite deleted blob",
+                exception);
+          } else if (result != null) {
+            // result can be null if the blob is not a composite blob.
+            for (StoreKey key : result) {
+              currentOperationsCount.incrementAndGet(); // @todo
+              currentBackgroundOperationsCount.incrementAndGet();
+              deleteManager.submitDeleteBlobOperation(key.getID(), new FutureResult<Void>(), new Callback<Void>() {
+                @Override
+                public void onCompletion(Void result, Exception exception) {
+                  if (exception != null) {
+                    logger.info("Background delete operation failed with exception", exception);
+                  }
+                  currentBackgroundOperationsCount.decrementAndGet();
+                }
+              });
+            }
+          }
+          currentBackgroundOperationsCount.decrementAndGet();
+        }
+      };
+      currentOperationsCount.incrementAndGet();
+      currentBackgroundOperationsCount.incrementAndGet();
+      getManager.submitGetChunkIdsOperation(blobId, futureResult, callback);
```

Comment:
Got a bit lost, I am wondering which part allows to get a blob after the blob is deleted?